conseil-api/src/main/scala/tech/cryptonomic/conseil/api/ConseilApi.scala

```diff
@@ -0,0 +1,136 @@
+
+  lazy val route: Route = {
+    lazy val platformDiscovery = PlatformDiscovery(metadataService)
+
+    lazy val routeDirectives = new RecordingDirectives
+    lazy val securityDirectives = new ValidatingDirectives(config.security)
+
+    import routeDirectives._
+    import securityDirectives._
+
+    concat(
+      pathPrefix("docs") {
+        pathEndOrSingleSlash {
+          getFromResource("web/index.html")
+        }
+      },
+      pathPrefix("swagger-ui") {
+        getFromResourceDirectory("web/swagger-ui/")
+      },
+      Docs.route,
+      cors() {
+        enableCORS {
+          implicit val correlationId: UUID = UUID.randomUUID()
+          handleExceptions(loggingExceptionHandler) {
+            extractClientIP {
+              ip =>
+                recordResponseValues(ip)(mat, correlationId) {
+                  timeoutHandler {
+                    concat(
+                      validateApiKey { _ =>
+                        concat(
+                          logRequest("Conseil", Logging.DebugLevel) {
+                            AppInfo.route
+                          },
+                          logRequest("Metadata Route", Logging.DebugLevel) {
+                            platformDiscovery.route
+                          },
+                          apiCache.cachedDataEndpoints(metadataService)
+                            .map {
+                              case (platform, routes) =>
+                                logRequest(s"$platform Data Route", Logging.DebugLevel) {
+                                  routes.getRoute ~ routes.postRoute
+                                }
+                            }
+                            .reduce(_ ~ _)
+                        )
+                      },
+                      options {
+                        // Support for CORS pre-flight checks.
+                        complete("Supported methods : GET and POST.")
+                      }
+                    )
+                  }
+                }
+            }
+          }
+        }
+      }
+    )
+  }
+
+  private class ApiCache {
+    private lazy val cache: Map[BlockchainPlatform, Api] = forVisiblePlatforms {
+      case Platforms.Tezos => new TezosApi(config.metadata, config.server)
+    }
+
+    lazy val cachedDiscoveryOperations: Map[String, PlatformDiscoveryOperations] = cache.map {
+      case (key, value) => key.name -> value.discoveryOperations
+    }
+
+    lazy val cachedDataEndpoints: MetadataService => Map[String, ApiDataEndpoints] = service =>
+      cache.map {
+        case (key, value) => key.name -> value.dataEndpoint(service)
+      }
+
+    private def forVisiblePlatforms(f: BlockchainPlatform => Api): Map[BlockchainPlatform, Api] =
+      transformation
+        .overridePlatforms(config.platforms.getPlatforms)
+        .map(toBlockchainPlatform)
+        .map(platform => platform -> f(platform))
```

Comment:
why two splitted `.map` calls and not a single

```scala
.map {
  p =>
    val platform = BlockchainPlatform.fromString(p.name)
    platform -> f(platform)
}
```

Reply:
Visibility mostly, I believe that this double `.map` would be optimized anyway by compiler (?)

Reply:
In this case I think so, but it's not always guaranteed. Especially with future composition...

I personally tend to avoid multiple map / flatmap calls that could be manually fused, but it's not a general rule, so I'd say it's ok.

Thanks for explaining